### ISSUE TITLE
Generated code not run when `outputModule: true` and entry chunk dependent on common chunk

### ISSUE THREAD

**Issue by ahabhgk** [webpack/webpack#17014]
# Bug report

**If the current behavior is a bug, please provide the steps to reproduce.**

repro #1: https://github.com/ahabhgk/webpack-output-module-repro
repro #2: https://github.com/webpack/webpack/issues/17870
repro #3: https://github.com/webpack/webpack/issues/13852

**What is the current behavior?**

index.js:
```js
import common from './common.js`
console.log(common)
```
common.js:
```js
export default true
```

webpack config:
```js
module.exports = {
  experiments: {
    outputModule: true,
  },
  target: 'es2020',
  optimization: {
    splitChunks: {
      cacheGroups: {
        common: {
          test: /common\.js/,
          minSize: 0,
          chunks: 'all'
        }
      }
    }
  }
}
```

run the build output: `node ./dist/main.mjs`, it will log nothing, but it should log `true`

**Other relevant information:**
webpack version:  ^5.79.0
Node.js version:  16
Operating System:  OSX

**Comment by alexander-akait** [comment]
Yeah, I see

**Comment by alexander-akait** [comment]
@ahabhgk Oh, there is not problems with webpack, something wrong with `target`, if you change it to `target: "web"` (or `node`) all works fine, but if you change it to `['es2020', 'web']`, it doesn't work

**Comment by alexander-akait** [comment]
Also if you speficity options manually (chunkFormat and etc) all works fine too, so something wrong with our defaults...

**Comment by alexander-akait** [comment]
Note: `target: 'es2020',` is not valid good value, because we don't know your env, so you should use `target: ['es2020', 'web']` or `['es2020', 'node']`, anyway if you use just `es2020` you will need to setup chunkFormat (and etc) manually

**Comment by ahabhgk** [comment]
I'm trying to find a chunkFormat that can run both on browser and node, [`target: "universal"`](https://github.com/webpack/webpack/issues/6525) would be perfect, but not implement yet, so I tried `chunkFormat: "module"` and found this bug.

**Comment by steverep** [comment]
Certainly should keep this open.  It also causes ESM web workers not to load if they are split using default optimization as discussed in #17298.

**Comment by althaf101** [comment]
@alexander-akait 
>https://github.com/webpack/webpack/issues/17014#issuecomment-1513226276

Could you please help with the config options that we need to pass manually until this is fixed ?  I tried the following. But didnt work !
```
  output: {
    ...
    chunkFormat: "module",
    chunkLoading: "import"
    asyncChunks: true
  },
```

**Comment by alexander-akait** [comment]
bump

**Comment by runspired** [comment]
I think I am also hitting this but without `outputModule: true` being involved:

I have two entrypoints (A, B) depending on a shared module (C)

A both imports from C and re-exports several of the things in C.

In dev, everything works. In prod, webpack only re-exports the things from C, while additional code is still present in A its own exports are not exported.

**Comment by alexander-akait** [comment]
@ahabhgk I think the same - https://github.com/webpack/webpack/issues/13852, did you find a fix in rspack?

**Comment by ahabhgk** [comment]
No, I haven't found a suitable fix that can be built into Rspack/or webpack, we usually just let user to use `chunks: "async"` and disable runtime chunk to avoid split the entry chunk

**Comment by alexander-akait** [comment]
@ahabhgk Thanks for the feedback, I'll try to find a solution (what's the problem) after the next issue
